src/networking.c

```diff
@@ -42,7 +42,7 @@ static void setProtocolError(const char *errstr, client *c);
 static void pauseClientsByClient(mstime_t end, int isPauseClientAll);
 int postponeClientRead(client *c);
 char *getClientSockname(client *c);
-int ProcessingEventsWhileBlocked = 0; /* See processEventsWhileBlocked(). */
```

Comment:
start version: 6.2.0
introduced: #8141
reason: release client argv outside of GIL

```
WARNING: ThreadSanitizer: data race (pid=1969717)
  Write of size 4 at 0x558d01163de0 by thread T14 (mutexes: write M64):
    #0 processEventsWhileBlocked /data/redis_fork/src/networking.c:4116 (redis-server+0xe795d)
    #1 RM_Yield /data/redis_fork/src/module.c:2410 (redis-server+0x20f4c3)
    #2 RM_Yield /data/redis_fork/src/module.c:2381 (redis-server+0x20f4c3)
    #3 bg_call_worker /data/redis_fork/tests/modules/blockedclient.c:116 (blockedclient.so+0x9ade)

  Previous read of size 4 at 0x558d01163de0 by main thread (mutexes: write M86, write M82, write M84):
    #0 afterSleep /data/redis_fork/src/server.c:1867 (redis-server+0x9f4ab)
    #1 aeProcessEvents /data/redis_fork/src/ae.c:408 (redis-server+0x971dc)
    #2 aeMain /data/redis_fork/src/ae.c:496 (redis-server+0x971dc)
    #3 main /data/redis_fork/src/server.c:7212 (redis-server+0x846d5)
```

Reply:
so we have one mutex protecting the times in which we do actual work (our GIL), and another mutex for protecting times when we go to sleep (not to use epoll in 2 threads at the same time)?

if we documented that using the RM_Yield API from a thread has to be done while locked, and it also attempts to lock the AE mutex, then it can only actually run in a very narrow time window. i wonder if that's the right approach.
maybe instead we can block the module thread, and wait for the main tread to perform an AE cycle?

i wonder why at all we allowed it to be run from a thread. i don't remember the use case.
@MeirShpilraien maybe you remember something?

Reply:
I discussed this with Meir.
Context: calling RM_Yield from a thread is meant to be done in case you locked the GIL for a long period of time, and don't intend to release it, and meanwhile, want to reply with -BUSY.
In this case, it is likely that that the main thread is already blocked trying to lock the GIL.
However, if server.hz is really low, and there are no clients with pending commands, we can imagine that the main thread's event loop is never gonna wake up, in which case RM_Yield will hang (despite wishing to use AE_DONT_WAIT).

The alternative solution we wanna propose is that instead of adding the new mutex to guard the AE mechanism, we'll change the code so that we process these events from the main thread, instead of the module thread.
1. the module sets server.busy_module_yield_flags and server.busy_module_yield_reply.
2. the module sends a character on the pipe to cause the main thread's event loop to wake up.
3. the module unlocks the GIL, waits for redis to reach the event loop once, and then re-locks the GIL.

p.s. if we agree on the direction, maybe we should extract this topic to another PR?

Reply:
It seems to be working, I'll give it a shot.

Reply:
@oranagra We're missing the fact that `RM_Yield()` can be called in the main thread, and if we want to proceed in that direction, we'll need to use old behavior in the main thread, and new direction in the module thread.

Reply:
ok, let's do that.
this brings back the detection of which thread we're running from, which is something we weren't happy with, but still i guess that's the right way forward.